File: app/src/lib/automod.ts
/**
 * AutoMod côté rendu (modèle serverless) : les clients honnêtes masquent les
 * mots filtrés du groupe à l'affichage — rien n'est supprimé du réseau. La
 * correspondance est insensible à la casse ET aux accents, et par mot entier
 * (frontières Unicode : lettres, chiffres et `_` sont des caractères de mot),
 * donc un mot filtré au milieu d'un autre mot n'est pas masqué.
 *
 * 🔗 JUMEAU de `crates/accord-core/src/automod.rs`, qui applique la MÊME règle
 * pour retrancher les messages masqués du compteur de non-lus. Les deux
 * doivent rester d'accord : toute évolution ici se reporte là-bas, et les deux
 * suites de tests partagent volontairement les mêmes cas (« concert » contre
 * « con », accents précomposés ou décomposés).
 */

/** Caractère de masquage affiché à la place d'un mot filtré. */
const MASK_CHAR = '█';

/** Longueur minimale du masque (ne pas révéler les mots très courts). */
const MASK_MIN = 3;

/** Longueur maximale du masque (ne pas révéler les mots très longs). */
const MASK_MAX = 8;

/**
 * Nombre maximal de mots filtrés par serveur.
 *
 * 🔒 C'est la borne du NŒUD (`MAX_AUTOMOD_WORDS`, `accord-proto`), pas une
 * borne d'interface : le format filaire refuse une liste plus longue, et
 * `groups.automod.set` la rejette avant d'émettre l'op. L'écran de réglage
 * affichait 100 alors que le nœud en acceptait 50 — au-delà du cinquantième
 * mot, « Enregistrer » échouait sur une erreur que rien n'annonçait.
 */
export const MAX_AUTOMOD_WORDS = 50;

/** Longueur maximale d'un mot filtré, en caractères (borne du nœud). */
export const MAX_AUTOMOD_WORD_CHARS = 32;

/**
 * Classe des caractères « de mot » pour les frontières Unicode.
 *
 * `Alphabetic` (et non `L`) parce que c'est exactement ce que rend
 * `char::is_alphabetic` côté Rust : le jumeau `accord-core/src/automod.rs` doit
 * placer les frontières aux mêmes endroits, sinon un message masqué ici
 * resterait compté dans la pastille de non-lus, ou l'inverse.
 */
const WORD_CHAR = '[\\p{Alphabetic}\\p{N}_]';

/**
 * Marques diacritiques combinantes, retirées au repli.
 *
 * Le même mot arrive tantôt précomposé (« é » = U+00E9), tantôt décomposé
 * (« e » + U+0301) selon le clavier, le système ou le copier-coller de
 * l'émetteur — macOS produit couramment la forme décomposée. Sans ce repli,
 * le filtre marchait ou non selon la machine d'en face, ce qui est pire
 * qu'un filtre absent : il donne l'illusion de protéger.
 */
const COMBINING = /[\u0300-\u036f]/g;

/**
 * Texte replié (minuscules, sans diacritiques) et correspondance d'index vers
 * le texte d'origine.
 *
 * Le repli ne conserve PAS les longueurs (« É » décomposé occupe deux unités
 * de code et se replie sur une), donc masquer aux offsets du texte replié
 * découperait le texte d'origine au mauvais endroit. `offsets[i]` donne
 * l'offset d'origine de la i-ème unité de code repliée ; la dernière entrée
 * est une sentinelle qui borne une correspondance allant jusqu'au bout.
 */
interface Folded {
  text: string;
  offsets: number[];
}

/** Replie un texte et construit sa correspondance d'index. */
function fold(input: string): Folded {
  let text = '';
  const offsets: number[] = [];
  let at = 0;
  // Itération par POINT de code (et non par unité de code) : un caractère
  // hors BMP ne doit pas être coupé en deux moitiés de paire de substitution.
  for (const ch of input) {
    const piece = ch.toLowerCase().normalize('NFD').replace(COMBINING, '');
    for (let i = 0; i < piece.length; i++) offsets.push(at);
    text += piece;
    at += ch.length;
  }
  offsets.push(input.length);
  return { text, offsets };
}

/** Échappe les métacaractères d'expression régulière d'un mot filtré. */
function escapeRegExp(word: string): string {
  return word.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Expression d'un mot filtré DÉJÀ replié : occurrence entière (pas de
 * caractère de mot immédiatement avant ni après), Unicode, globale. La casse
 * et les accents sont déjà neutralisés par [`fold`], d'où l'absence du
 * drapeau `i`.
 */
function wordPattern(foldedWord: string): RegExp {
  return new RegExp(`(?<!${WORD_CHAR})${escapeRegExp(foldedWord)}(?!${WORD_CHAR})`, 'gu');
}

/** Mots exploitables d'une liste AutoMod, repliés ; les vides sont écartés. */
function usableWords(words: readonly string[]): string[] {
  return words.map((w) => fold(w.trim()).text).filter((w) => w !== '');
}

/**
 * Occurrence d'un mot filtré : bornes `[from, to[` dans le texte d'ORIGINE
 * (pour découper au bon endroit) et longueur `seen` du mot une fois replié
 * (pour dimensionner le masque sans compter les accents décomposés deux fois).
 */
type Occurrence = { from: number; to: number; seen: number };

/** Occurrences des mots filtrés, triées et sans chevauchement. */
function filteredRanges(text: string, words: readonly string[]): Occurrence[] {
  const useful = usableWords(words);
  if (useful.length === 0) return [];
  const { text: hay, offsets } = fold(text);
  const found: Occurrence[] = [];
  for (const word of useful) {
    for (const m of hay.matchAll(wordPattern(word))) {
      found.push({
        from: offsets[m.index] ?? text.length,
        to: offsets[m.index + m[0].length] ?? text.length,
        seen: [...m[0]].length,
      });
    }
  }
  found.sort((a, b) => a.from - b.from);
  // Deux mots filtrés peuvent se recouvrir (« con » et « concon ») : garder le
  // premier suffit, le second est déjà sous le masque.
  const merged: Occurrence[] = [];
  let last = -1;
  for (const occ of found) {
    if (occ.from < last) continue;
    merged.push(occ);
    last = occ.to;
  }
  return merged;
}

/**
 * Masque chaque occurrence (mot entier, insensible à la casse et aux accents)
 * des mots filtrés par des `█` de longueur bornée ([3, 8]) proche de celle du
 * mot.
 */
export function maskFiltered(text: string, words: readonly string[]): string {
  const ranges = filteredRanges(text, words);
  if (ranges.length === 0) return text;
  let out = '';
  let cursor = 0;
  for (const { from, to, seen } of ranges) {
    out += text.slice(cursor, from);
    out += MASK_CHAR.repeat(Math.min(MASK_MAX, Math.max(MASK_MIN, seen)));
    cursor = to;
  }
  return out + text.slice(cursor);
}

/**
 * Vrai si `text` contient au moins un mot filtré (même règle de
 * correspondance que [`maskFiltered`]) — pour l'avertissement émetteur et
 * pour couper notification et son d'un message masqué.
 */
export function containsFiltered(text: string, words: readonly string[]): boolean {
  const useful = usableWords(words);
  if (useful.length === 0) return false;
  const { text: hay } = fold(text);
  return useful.some((word) => wordPattern(word).test(hay));
}

File: crates/accord-core/src/automod.rs
//! Correspondance AutoMod côté nœud (SPEC §7, modèle sans serveur).
//!
//! 🔗 JUMEAU de `app/src/lib/automod.ts`. L'interface masque les mots filtrés
//! au rendu ; le nœud applique ici la MÊME règle pour retrancher les messages
//! masqués du compteur de non-lus. Les deux doivent décider pareil : un
//! message dont le mot est remplacé par des `█` mais qui continue d'allumer la
//! pastille rouge désigne exactement ce que le filtre prétendait cacher.
//!
//! Les deux implémentations sont volontairement écrites à l'identique :
//! minuscules, décomposition canonique, suppression des diacritiques
//! combinantes, puis recherche d'occurrence bornée aux frontières de mot
//! (`Alphabetic`, `N` et `_` sont des caractères de mot des deux côtés). Les
//! suites de tests partagent les mêmes cas, nommés pareil.
//!
//! Ce module ne décide RIEN : il ne supprime pas, ne réécrit pas, ne bloque
//! aucune émission. L'AutoMod d'Accord est une convention d'affichage entre
//! clients honnêtes, la liste de mots voyage en clair dans l'op-log signée du
//! groupe, et un client modifié verra toujours le texte entier.

use accord_proto::core_msg::MsgBody;
use unicode_normalization::UnicodeNormalization;

/// Vrai si `c` fait partie d'un mot : ce sont ces caractères qui, collés au
/// mot filtré, empêchent la correspondance (« concert » contre « con »).
///
/// `is_alphabetic() || is_numeric()` correspond exactement à
/// `[\p{Alphabetic}\p{N}]` du jumeau TypeScript.
fn is_word_char(c: char) -> bool {
    c.is_alphabetic() || c.is_numeric() || c == '_'
}

/// Forme comparable d'un texte ou d'un mot filtré : minuscules, décomposé,
/// diacritiques combinantes retirées.
///
/// Le même mot arrive tantôt précomposé (« é » = U+00E9), tantôt décomposé
/// (« e » + U+0301) selon le clavier ou le système de l'émetteur — macOS
/// produit couramment la forme décomposée. Sans ce repli, le filtre marchait
/// ou non selon la machine d'en face.
fn fold(input: &str) -> String {
    input
        .chars()
        .flat_map(char::to_lowercase)
        .nfd()
        .filter(|c| !('\u{0300}'..='\u{036F}').contains(c))
        .collect()
}

/// Prépare la liste de mots filtrés d'un groupe : repli une fois pour toutes,
/// mots vides écartés.
///
/// Séparée de [`matches`] parce qu'un salon compare la MÊME liste à des
/// centaines de messages : la replier à chaque message serait du travail
/// refait pour rien.
pub fn prepare<S: AsRef<str>>(words: impl IntoIterator<Item = S>) -> Vec<String> {
    words
        .into_iter()
        .map(|w| fold(w.as_ref().trim()))
        .filter(|w| !w.is_empty())
        .collect()
}

/// Vrai si `hay` contient `needle` comme mot entier. Les deux sont attendus
/// déjà repliés par [`fold`].
fn contains_word(hay: &str, needle: &str) -> bool {
    if needle.is_empty() {
        return false;
    }
    let mut from = 0;
    while let Some(rel) = hay[from..].find(needle) {
        let at = from + rel;
        let end = at + needle.len();
        let left_ok = hay[..at]
            .chars()
            .next_back()
            .is_none_or(|c| !is_word_char(c));
        let right_ok = hay[end..].chars().next().is_none_or(|c| !is_word_char(c));
        if left_ok && right_ok {
            return true;
        }
        // Avance d'UN caractère : `hay[at..]` commence par `needle`, donc la
        // longueur du premier caractère de `needle` est une frontière valide
        // (un `+ 1` brut couperait un caractère multi-octets et paniquerait).
        from = at + needle.chars().next().map_or(1, char::len_utf8);
    }
    false
}

/// Vrai si `text` contient au moins un mot filtré. `prepared` sort de
/// [`prepare`].
pub fn matches(text: &str, prepared: &[String]) -> bool {
    if prepared.is_empty() || text.is_empty() {
        return false;
    }
    let hay = fold(text);
    prepared.iter().any(|w| contains_word(&hay, w))
}

/// Vrai si le message est masqué par l'AutoMod, d'après le texte que
/// l'interface AFFICHERAIT.
///
/// Même choix que `displayText` côté interface : la dernière édition prime sur
/// le corps d'origine, et seul un corps texte porte du texte masquable. Les
/// sondages, stickers et pièces jointes ne sont pas masqués au rendu
/// (`PollCard`, `StickerImage`, `AttachmentRow` ignorent la liste) ; les
/// compter comme masqués ferait retomber une pastille sur un message
/// parfaitement lisible.
pub fn message_matches(kind: u8, body: &[u8], edited: Option<&[u8]>, prepared: &[String]) -> bool {
    if prepared.is_empty() {
        return false;
    }
    if let Some(edit) = edited {
        return matches(&String::from_utf8_lossy(edit), prepared);
    }
    match MsgBody::decode_body(kind, body) {
        Ok(MsgBody::Text { text, .. }) => matches(&text, prepared),
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn kw(words: &[&str]) -> Vec<String> {
        prepare(words)
    }

    // jumeau: « ne masque pas un mot filtré au milieu d'un autre mot »
    #[test]
    fn matches_whole_words_only() {
        let words = kw(&["con"]);
        assert!(matches("quel con celui-là", &words));
        // Le cas qui donne son sens à la frontière de mot.
        assert!(!matches("on va au concert", &words));
        assert!(!matches("un jambon suffit", &words));
        assert!(matches("con.", &words));
        assert!(matches("(con)", &words));
        assert!(matches("con", &words));

        let chat = kw(&["chat"]);
        assert!(!matches("le chaton dort", &chat));
        assert!(!matches("achat en ligne", &chat));
        assert!(matches("le chat et le chaton", &chat));
        // Frontière Unicode : une lettre accentuée collée reste le même mot.
        assert!(!matches("idiotè", &kw(&["idiot"])));
    }

    #[test]
    fn matches_ignores_case() {
        let words = kw(&["idiot"]);
        assert!(matches("IDIOT", &words));
        assert!(matches("Idiot", &words));
        assert!(matches("quel IDIOT alors", &words));
    }

    // jumeau: « masque un mot accentué filtré sans accent, et l'inverse »
    #[test]
    fn matches_ignores_accents_in_both_directions() {
        assert!(matches("espèce de crétin", &kw(&["cretin"])));
        assert!(matches("espece de cretin", &kw(&["crétin"])));
        assert!(matches("CRÉTIN va", &kw(&["crétin"])));
    }

    // jumeau: « masque quel que soit le codage Unicode de l'accent (NFC/NFD) »
    #[test]
    fn matches_precomposed_and_decomposed_alike() {
        let nfc = "crétin";
        let nfd = "cre\u{0301}tin";
        assert_ne!(nfc, nfd, "les deux formes doivent bien différer en octets");
        assert!(matches(&format!("quel {nfd} !"), &kw(&[nfc])));
        assert!(matches(&format!("quel {nfc} !"), &kw(&[nfd])));
    }

    #[test]
    fn matches_handles_empty_inputs() {
        assert!(!matches("n'importe quoi", &[]));
        assert!(!matches("", &kw(&["con"])));
        // Un mot vide ou blanc est écarté par `prepare`, jamais comparé.
        assert!(kw(&["", "   "]).is_empty());
        assert!(!matches("n'importe quoi", &kw(&["   "])));
    }

    #[test]
    fn matches_does_not_stop_at_the_first_glued_occurrence() {
        // « concert » vient AVANT le mot isolé : la recherche doit continuer.
        assert!(matches("le concert puis con", &kw(&["con"])));
    }

    #[test]
    fn message_matches_reads_text_and_edits_only() {
        let words = kw(&["spoiler"]);
        let text = MsgBody::Text {
            text: "attention spoiler".into(),
            reply_to: None,
            attachments: vec![],
        };
        assert!(message_matches(
            text.kind(),
            &text.encode_body(),
            None,
            &words
        ));

        let clean = MsgBody::Text {
            text: "rien à signaler".into(),
            reply_to: None,
            attachments: vec![],
        };
        assert!(!message_matches(
            clean.kind(),
            &clean.encode_body(),
            None,
            &words
        ));
        // Édité POUR contenir le mot : c'est le texte AFFICHÉ qui compte, et
        // c'est bien lui que l'interface masque.
        assert!(message_matches(
            clean.kind(),
            &clean.encode_body(),
            Some("finalement spoiler".as_bytes()),
            &words
        ));
        // ... et l'inverse : édité pour ne plus le contenir.
        assert!(!message_matches(
            text.kind(),
            &text.encode_body(),
            Some("rien finalement".as_bytes()),
            &words
        ));

        // Un sondage n'est pas masqué au rendu : il ne doit pas non plus
        // disparaître du compteur de non-lus.
        let poll = MsgBody::Poll {
            poll_id: [3; 16],
            question: "quel spoiler ?".into(),
            options: vec!["oui".into(), "non".into()],
        };
        assert!(!message_matches(
            poll.kind(),
            &poll.encode_body(),
            None,
            &words
        ));

        // Sans mot filtré, aucun corps n'est examiné.
        assert!(!message_matches(
            text.kind(),
            &text.encode_body(),
            None,
            &[]
        ));
        // Corps indécodable : jamais masqué.
        assert!(!message_matches(200, b"\xff\xff", None, &words));
    }
}
